src/main/java/org/codehaus/plexus/components/io/attributes/FileAttributes.java

```diff
@@ -52,69 +53,94 @@
-
-            groupId = (Integer) attrs.get( "gid" );
-
-            String groupName = groupCache.get( groupId );
-            if ( groupName != null )
-            {
-                this.groupName = groupName;
-            }
-            else
-            {
-                Object group = Files.getAttribute( path, "unix:group", LinkOption.NOFOLLOW_LINKS );
-                this.groupName = ( (Principal) group ).getName();
-                groupCache.put( groupId, this.groupName );
-            }
-            userId = (Integer) attrs.get( "uid" );
-            String userName = userCache.get( userId );
-            if ( userName != null )
-            {
-                this.userName = userName;
-            }
-            else
-            {
-                Object owner = Files.getAttribute( path, "unix:owner", LinkOption.NOFOLLOW_LINKS );
-                this.userName = ( (Principal) owner ).getName();
-                userCache.put( userId, this.userName );
-            }
-            octalMode = (Integer) attrs.get( "mode" ) & 0xfff; // Mask off top bits for compatibilty. Maybe check if we
-                                                               // can skip this
-            symbolicLink = (Boolean) attrs.get( "isSymbolicLink" );
+            names = "unix:*";
+        }
+        else if ( views.contains( "posix" ) )
+        {
+            names = "posix:*";
         }
         else
         {
-            FileOwnerAttributeView fa = AttributeUtils.getFileOwnershipInfo( file );
-            this.userName = fa.getOwner().getName();
-            userId = null;
-            this.groupName = null;
-            this.groupId = null;
-            octalMode = PlexusIoResourceAttributes.UNKNOWN_OCTAL_MODE;
-            permissions = Collections.emptySet();
-            symbolicLink = Files.isSymbolicLink( path );
+            names = "basic:*";
+        }
+        Map<String, Object> attrs = Files.readAttributes( path, names, LinkOption.NOFOLLOW_LINKS);
+        if ( !attrs.containsKey( "group" ) && !attrs.containsKey( "owner" ) && views.contains( "owner" ) )
+        {
+            Map<String, Object> ownerAttrs = Files.readAttributes( path, "owner:*", LinkOption.NOFOLLOW_LINKS);
+            Map<String, Object> newAttrs = new HashMap<>( attrs );
+            newAttrs.putAll( ownerAttrs );
+            attrs = newAttrs;
         }
+        this.groupId = (Integer) attrs.get( "gid" );
+        this.groupName = attrs.containsKey( "group" ) ? ((Principal) attrs.get( "group" ) ).getName() : null;
+        this.userId = (Integer) attrs.get( "uid" );
+        this.userName = attrs.containsKey( "owner" ) ? ((Principal) attrs.get( "owner" ) ).getName() : null;
+        this.symbolicLink = (Boolean) attrs.get( "isSymbolicLink" );
+        this.regularFile = (Boolean) attrs.get( "isRegularFile" );
+        this.directory = (Boolean) attrs.get( "isDirectory" );
+        this.other = (Boolean) attrs.get( "isOther" );
```

Comment:
Can those be null and end up in `NPE`?

Reply:
Those are basic attributes and supposed to be always provided by the file system implementation.  The only basic attributed that is optional is the `fileKey`.

Reply:
Alright, thanks!